### PR TITLE
Tickets/dm 35483

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

File: src/partition/sph-partition-matches.cc [hunk 1]
```diff
@@ -160,54 +217,124 @@ Worker::Worker(ConfigStore const& config)
...
-        }
-        if (!null2) {
-            sc2.first = _editor.get<double>(_pos2.first);
-            sc2.second = _editor.get<double>(_pos2.second);
-            loc2 = _chunker.locate(sc2);
-        }
-        if (!null1) {
-            _editor.set(_chunkIdField, loc1.chunkId);
-            _editor.set(_subChunkIdField, loc1.subChunkId);
+        if (_id1Field != -1) {
```
Comment (andy-slac) [inline src/partition/sph-partition-matches.cc:223]: This is a very long if/else with some code duplication and non-trivial control flow. It's hard to follow and verify that there is no mixup in variable names (sc1/sc2, loc1/loc2, etc). Some of these code could probably be moved to a separate method/class.

File: src/partition/tests/objectIndex.cc [hunk 2]
```diff
@@ -77,88 +77,104 @@ BOOST_AUTO_TEST_CASE(ObjectIndexTest) {
...
+        BOOST_CHECK_EQUAL(ObjectIndex::instance(indexName)->isOpen(), true);
+        BOOST_CHECK_EQUAL(ObjectIndex::instance(indexName)->mode(), ObjectIndex::READ);
+        std::pair<int32_t, int32_t> chunkSubChunk;
+        BOOST_REQUIRE_NO_THROW({ chunkSubChunk = ObjectIndex::instance(indexName)->read(validId); });
+        int32_t const chunkId = chunkSubChunk.first;
+        int32_t const subChunkId = chunkSubChunk.second;
+        BOOST_CHECK_EQUAL(chunkId, 1U);
+        BOOST_CHECK_EQUAL(subChunkId, 2U);
+    }
+}
```
Comment (andy-slac) [inline src/partition/tests/objectIndex.cc:180]: Newline is missing.

**Review by andy-slac — APPROVED** [review]
Looks OK, few minor comments.

File: src/partition/ChunkReducer.cc [hunk 3]
```diff
@@ -91,7 +91,7 @@ void ChunkReducer::_makeFilePaths(int32_t chunkId) {
     if (_numNodes > 1) {
         // Files go into a node-specific sub-directory.
         char subdir[32];
-        uint32_t node = hash(static_cast<uint32_t>(chunkId)) % _numNodes;
+        uint32_t node = std::hash<uint32_t>{}(static_cast<uint32_t>(chunkId)) % _numNodes;
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/partition/ChunkReducer.cc:94]: Good point. As a side comment, I don't understand why chunk numbers are represented by the signed integers in Qserv code. These numbers can't be negative.

File: src/partition/ObjectIndex.h [hunk 4]
```diff
@@ -83,13 +83,17 @@ class ObjectIndex {
...
      */
-    static ObjectIndex& instance() {
-        static ObjectIndex oi;
-        return oi;
+    static ObjectIndex* instance(std::string const& name = std::string()) {
+        static std::mutex mtx;
+        std::lock_guard<std::mutex> lock(mtx);
+        static std::map<std::string, std::unique_ptr<ObjectIndex>> oi;
+        if (auto itr = oi.find(name); itr == oi.end()) oi[name].reset(new ObjectIndex());
+        return oi[name].get();
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/partition/ObjectIndex.h:96]: Agreed with the first suggestion (to optimize the implementation of the factory method). As far as the second one is concerned, a problem here is the mini-framework that won't let me create indexes and pass them from `main()` into the template classes of the framework:
```
typedef Job<Worker> PartitionMatchesJob;
```
I'm not ready (yet) for redesigning this code. I will do it  next time I will have to make major changes here.

File: src/partition/sph-partition-matches.cc [hunk 5]
```diff
@@ -160,54 +217,124 @@ Worker::Worker(ConfigStore const& config)
...
+                int32_t const chunkId1 = chunkSubChunk1.first;
+                int32_t const subChunkId1 = chunkSubChunk1.second;
+                _editor.set(_chunkIdField, chunkId1);
+                _editor.set(_subChunkIdField, subChunkId1);
+                if (!null2) {
+                    // Both positions are valid.
+                    int32_t const chunkId2 = chunkSubChunk2.first;
+                    if (chunkId1 == chunkId2) {
+                        // Both positions are in the same chunk.
+                        _editor.set(_flagsField, '3');
```
[the remark this answers is not in the document]
Comment (fritzm) [reply src/partition/sph-partition-matches.cc:270]: Yeah, I had a look for that the other day, because the meaning/use of the flag column was unfamiliar to me.  I didn't see anywhere where it was actually used/referenced beyond the partitioner.  I think it is likely dead wood.
